ptable/ptable.c

```diff
@@ -0,0 +1,122 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Partition table
+ *
+ * Copyright 2020 Phoenix Systems
+ * Author: Hubert Buczynski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <string.h>
+#include <stdlib.h>
+#include <ctype.h>
+#include <errno.h>
+
+#include "ptable.h"
+
+
+#define MIN_PARTITIONS_SECTORS_NB    3
+
+
+static int ptable_checkType(uint8_t type)
+{
+	int res = EOK;
+
+	switch (type) {
+		case ptable_raw:
+		case ptable_meterfs:
+			break;
+
+		default:
+			res = -EINVAL;
+			break;
+	}
+
+	return res;
+}
+
+
+int ptable_verifyPartition(uint8_t id, const ptable_partition_t *pHeaders, const memory_properties_t *mem)
+{
+	int i;
+
+	/* Verify size and offset */
+	if (pHeaders[id].size % mem->sectorSize)
+		return -EINVAL;
+
+	if (pHeaders[id].offset % mem->sectorSize)
+		return -EINVAL;
+
+	if ((pHeaders[id].size + pHeaders[id].offset) > mem->memSize)
+		return -EINVAL;
+
+	if (pHeaders[id].size < MIN_PARTITIONS_SECTORS_NB * mem->sectorSize)
+		return -EINVAL;
+
+	/* Check partitions overlapping according to order in partition table */
+	for (i = 0; i < id; ++i) {
+		if (((pHeaders[id].offset > pHeaders[i].offset) && (pHeaders[id].offset < pHeaders[i].offset + pHeaders[i].size)) ||
+			((pHeaders[id].offset + pHeaders[id].size > pHeaders[i].offset) && (pHeaders[id].offset + pHeaders[id].size < pHeaders[i].offset + pHeaders[i].size)))
+			return -EINVAL;
+	}
+
+	/* Verify partition type */
+	if (ptable_checkType(pHeaders[id].type) < 0)
+		return -EINVAL;
+
+	/* Check partition name */
+	for (i = 0; i < strlen((const char *)pHeaders[id].name); ++i) {
+		if (!isalnum(pHeaders[id].name[i]))
+			return -EINVAL;
+	}
+
+	return EOK;
+}
+
+
+int ptable_getPartitions(ptable_partition_t **pHeaders, uint8_t *pHeadersLen, const memory_properties_t *mem)
+{
+	uintptr_t ptabAddr;
+	uintptr_t partsAddr;
+	uintptr_t magicAddr;
+	uint16_t maxPartCnt;
```

Comment:
no gain in using `uint16_t`, in every expression below it will be promoted to at least uint32_t